clvm2clusterMD_migration_bsc#1123496

taroth21 opened the pull request:
Tech review appreciated.

taroth21 commented on xml/ha_cluster_lvm.xml:
```diff
@@ -862,7 +862,28 @@ vdc                     253:32   0   20G  0 disk</screen>
     </para>
     <screen>&prompt.root;mdadm --create /dev/md0 --bitmap=clustered \
-  --metadata=1.2 --raid-devices=1 --force --level=mirror /dev/vdc</screen>
+    --metadata=1.2 --raid-devices=1 --force --level=mirror --data-offset=??? /dev/vdc</screen>
```
Thanks for the input, Roger. I tried to sum up the conditions that need to be checked before migration in an 'Important' note and moved the details for ```--data-offset``` to a separate paragraph within the step.

taroth21 commented on xml/ha_cluster_lvm.xml:
```diff
@@ -862,7 +862,28 @@ vdc                     253:32   0   20G  0 disk</screen>
+    -->
+    <para>
+     When <command>mdadm</command> creates an array, it reserves space at the
+     start of devices, which can be used for reshaping (re-striping) the array.
```
Added this information.

taroth21 commented:
Here's a draft PDF of the chapter with the latest changes:
[cha.ha.clvm_remarks_color_draft_en.pdf](https://github.com/SUSE/doc-sleha/files/2926615/cha.ha.clvm_remarks_color_draft_en.pdf)

zzhou1 commented:
> Here's a draft PDF of the chapter with the latest changes:
> [cha.ha.clvm_remarks_color_draft_en.pdf](https://github.com/SUSE/doc-sleha/files/2926615/cha.ha.clvm_remarks_color_draft_en.pdf)
> 
> Please review p.11-13 once more. Thanks.

Sounds good! 

Well, enjoying the reading of the draft pdf, I am thinking about user/admin's usual behaviors. They are more often just copy&paste those commands for their system. Hence, I changed my mind from the view as a "developer" to a "user", and vote to add --data-offset as the default in HA DOC. Technically, those padding space is useless for the current cluster md raid1/raid10, and we don't foresee the need to reshape cluster md raid5 for many years to come. The advantage is just to ease the use case for `--mirorlog mirrored` by default, and no harm to others at all.

taroth21 commented:
> Hence, I changed my mind from the view as a "developer" to a "user", and vote to add --data-offset as the default in HA DOC. Technically, those padding space is useless for the current cluster md raid1/raid10, and we don't foresee the need to reshape cluster md raid5 for many years to come. The advantage is just to ease the use case for `--mirorlog mirrored` by default, and no harm to others at all.

No problem, good point. :) I moved ```--data-offset``` back to the command in the step (and modified the 'important' note and the following para slightly to reflect this). Should we also include a value for the offset in the command (like ```--data-offset=128KB```, for example)? 
(Not sure which value would make sense, please also check the syntax and unit in which the offset needs to be entered (KB?)). Thanks!

zzhou1 commented:
> Should we also include a value for the offset in the command (like `--data-offset=128KB`, for example)?
> (Not sure which value would make sense, please also check the syntax and unit in which the offset needs to be entered (KB?)). Thanks!

`man mdadm` --data-offset
"Setting the offset explicitly over-rides the default.  The value given is in Kilobytes unless a suffix of 'K', 'M' or 'G' is used"

tomschr commented on xml/ha_cluster_lvm.xml:
```diff
@@ -812,17 +821,63 @@ vdc                                   253:32   0   20G  0 disk
+      During migration, reduce unused padding space from the cluster MD array
+      by running the <command>mdadm</command> command with
+      <option>--data-offset</option>. See <xref linkend="step.data-offset"/>
+      for details. By default, <command>mdadm</command> reserves 16 MB space at
```
Maybe add a `&nbsp;` between "16" and "MB" to avoid weird linebreaks? => `16&nbsp;MG`

taroth21 replied:
Done.

tomschr commented on xml/ha_cluster_lvm.xml:
```diff
@@ -856,13 +911,27 @@ vdb                     253:16   0   20G  0 disk
+    -->
+     The <option>--data-offset</option> option reduces unneeded padding space
+     from the array. The offset must be large enough for cluster MD to write its
+     data to it. We recommend 128 KB. On the other hand, the offset must be small
```
Same here with `&nbsp;` between "128" and "KB".

taroth21 replied:
Done.

tomschr commented on xml/ha_cluster_lvm.xml:
```diff
@@ -812,17 +821,63 @@ vdc                                   253:32   0   20G  0 disk
+      <option>--data-offset</option>. See <xref linkend="step.data-offset"/>
+      for details. By default, <command>mdadm</command> reserves 16 MB space at
+      the start of a device, which can be used for reshaping (re-striping) the
+      array. Reducing this space allows to fit in the LVM data and metadata.
```
Just to let you know one detail.

The 16MB is probably the superblock, right?

Although you mentioned "by default" and "at the start of the device", it may be a bit dangerous. The `mdadm` tool understands several metadata versions (0.90, 1.0, 1.1, 1.2, ...). According to the [mdadm manpage (option --metadata)](https://linux.die.net/man/8/mdadm):

>  The different sub-versions store the superblock at different locations on the device, either
>  at the end (for 1.0), at the start (for 1.1) or 4K from the start (for 1.2).  "1" is equivalent to
>  "1.2" (the commonly preferred 1.x format).  "default" is equivalent to "1.2".

If you refer really to the superblock, I would mention that:

    By default, <command>mdadm</command> reserves 16 MB space for the superblock...

Does that make sense?

taroth21 replied:
> The 16MB is probably the superblock, right?
> 
According to the [mdadm manpage (option --metadata)](https://linux.die.net/man/8/mdadm):
> 
> > The different sub-versions store the superblock at different locations on the device, either at the end (for 1.0), at the start (for 1.1) or 4K from the start (for 1.2).  "1" is equivalent to "1.2" (the commonly preferred 1.x format).  "default" is equivalent to "1.2".
> 
Very good question, I don't know (at least 'superblock' was not mentioned anywhere in the bug report). Handing over to @zzhou1 for this question.

tomschr commented on xml/ha_cluster_lvm.xml:
```diff
@@ -856,13 +911,27 @@ vdb                     253:16   0   20G  0 disk
+     volume extents of the migrated volume. Because the volume can have spanned
+     the complete device minus the mirror log, the offset must be smaller than
+     the size of the mirror log. If no value is specified for the offset,
+     its default value is 1 KB (1024 bytes).
```
Same here with `&nbsp;`.

taroth21 replied:
Done.

zzhou1 commented:
在 2019/3/8 22:16, Tanja Roth 写道:
>
>     The 16MB is probably the superblock, right?
>

No. Superblock is very small, eg. 4K.
[1](https://raid.wiki.kernel.org/index.php/RAID_superblock_formats)

--Roger

taroth21 commented:
> 在 2019/3/8 22:16, Tanja Roth 写道:
> The 16MB is probably the superblock, right?
> No. Superblock is very small, eg. 4K. [1](https://raid.wiki.kernel.org/index.php/RAID_superblock_formats)
> […](#)
> --Roger

Great, thanks for clarifying! In that case, I will ignore the comment by Thomas.
